producer/writer/retryable_connection.go

```diff
@@ -0,0 +1,259 @@
+		lastResetNanos:  0,
+		initialized:     false,
+		closed:          atomic.NewBool(false),
+		doneCh:          make(chan struct{}),
+		resetCh:         make(chan struct{}, 1),
+		m:               newRetryableConnectionMetrics(opts.InstrumentOptions().MetricsScope()),
+		nowFn:           time.Now,
+	}
+
+	c.connectFn = c.connectOnce
+	if err := c.tryResetWithConnectFn(c.connectFn); err != nil {
+		c.SignalReset()
+	}
+	return c
+}
+
+func (c *retryableConnection) Read(p []byte) (int, error) {
+	c.connLock.RLock()
+	if !c.initialized {
+		c.connLock.RUnlock()
+		return 0, errConnNotInitialized
+	}
+	n, err := c.r.Read(p)
+	c.connLock.RUnlock()
+	return n, err
+}
+
+func (c *retryableConnection) Write(p []byte) (int, error) {
+	c.connLock.RLock()
+	if !c.initialized {
+		c.connLock.RUnlock()
+		return 0, errConnNotInitialized
+	}
+	n, err := c.w.Write(p)
+	c.connLock.RUnlock()
+	return n, err
+}
+
+func (c *retryableConnection) SignalReset() {
+	select {
+	case c.resetCh <- struct{}{}:
+	default:
+	}
+}
+
+func (c *retryableConnection) Init() {
+	c.wg.Add(1)
+	go func() {
+		c.resetConnectionForever()
+		c.wg.Done()
+	}()
+}
+
+func (c *retryableConnection) resetConnectionForever() {
+	for {
+		select {
+		case <-c.resetCh:
+			c.tryResetWithConnectFn(c.connectWithRetry)
+		case <-c.doneCh:
+			if c.conn != nil {
+				c.conn.Close()
+			}
+			return
+		}
+	}
+}
+
+func (c *retryableConnection) tryResetWithConnectFn(fn connectFn) error {
+	// Avoid resetting too frequent.
+	if c.nowFn().UnixNano() < c.lastResetNanos+c.resetDelayNanos {
+		return nil
+	}
+	c.m.resetConn.Inc(1)
+	conn, err := fn(c.addr)
+	if err != nil {
+		c.m.resetError.Inc(1)
+		c.logger.Errorf("could not connect to %s, %v", c.addr, err)
+		return err
+	}
+	c.reset(conn)
+	return nil
+}
+
+func (c *retryableConnection) connectWithRetry(addr string) (net.Conn, error) {
+	continueFn := func(int) bool {
+		return !c.isClosed()
+	}
+	var (
+		conn net.Conn
+		err  error
+	)
+	fn := func() error {
+		conn, err = c.connectFn(addr)
+		return err
+	}
+	if attemptErr := c.retrier.AttemptWhile(
+		continueFn,
+		fn,
+	); attemptErr != nil {
+		return nil, fmt.Errorf("failed to connect to address %s, %v", addr, attemptErr)
+	}
+	return conn, nil
+}
+
+func (c *retryableConnection) connectOnce(addr string) (net.Conn, error) {
+	conn, err := net.DialTimeout("tcp", addr, c.opts.DialTimeout())
+	if err != nil {
+		c.m.connectError.Inc(1)
+		return nil, err
+	}
+	tcpConn := conn.(*net.TCPConn)
+	if err = tcpConn.SetKeepAlive(true); err != nil {
+		c.m.setKeepAliveError.Inc(1)
+	}
+	return tcpConn, nil
+}
+
+func (c *retryableConnection) reset(conn net.Conn) {
+	// Close the connection to wake up potential blocking write/read calls.
+	if c.conn != nil {
+		c.conn.Close()
+	}
+
+	c.connLock.Lock()
+	defer c.connLock.Unlock()
+
+	c.conn = conn
+	c.w.Reset(conn)
+	c.r.Reset(conn)
+	c.initialized = true
+	c.lastResetNanos = c.nowFn().UnixNano()
+	c.cleanUpResetChannel()
```

Comment:
Why is this needed?

Reply:
yeah seems I can remove this now since the delay check is moved to the reset thread